implementations/rust/ockam/ockam_api/src/nodes/service/node_services.rs

```diff
@@ -768,6 +768,31 @@ impl NodeManagerWorker {
         Ok(())
     }
 
+    pub(crate) async fn delete_kafka_service<'a>(
+        &'a self,
+        ctx: &Context,
+        req: &'a Request<'_>,
+        dec: &mut Decoder<'_>,
+        kind: KafkaServiceKind,
+    ) -> Result<ResponseBuilder> {
+        let body: DeleteServiceRequest = dec.decode()?;
+        let address = body.address();
+        let mut node_manager = self.node_manager.write().await;
+        let res = match node_manager.registry.kafka_services.remove(&address) {
+            None => Response::not_found(req.id()),
+            Some(e) => {
+                if !matches!(e.kind(), _kind) {
```

Comment:
Shouldn't you test for equality here since matching against a pattern with a variable will always succeed?
My other comment would be that it is a bit weird to accept the operation, do the deletion, and then return an error. We should better do the check before deleting the worker.

Reply:
Returning the address back is indeed one way to do it, but it feels a bit hacky to me. 
An extension method on the kafka_registry would be better IMO:

 - instead of calling `remove` you call `remove_consumer`
 - then the implementation of `remove_consumer` gets the info at the requested address and checks if this is really a consumer before removing it

That does not even have to be an extension method, just a get / check / delete (in case of a successful check) sequence, instead of delete / check / restore (in case of a failed check) sequence.

Reply:
> That does not even have to be an extension method, just a get / check / delete (in case of a successful check) sequence, instead of delete / check / restore (in case of a failed check) sequence.

- `get / check / delete`: implies two lookups in the best case scenario (we remove the entry); and one in the worst case
- `delete / check / restore`: the other way around (1 lookup in the best case, 2 in the worst)

In this case I opted for the second case because it favors the most common scenario (the user calls the right command to delete a given kafka service) at the expense of having a little weirder logic.

What are your thoughts on this? I know it's an unnecessary optimization, but doing it doesn't change the logic too much either...

Reply:
I definitely prefer the first kind of logic because it prevents inconsistencies in general. I don't think this matters tremendously in that specific case but in general it does. If something has been removed when it shouldn't have been other processes might rely on that thing to exist.

The other alternative is to have one registry for producers and one registry for consumers then there's nothing to check.

Reply:
Ok you convinced me 😆 Sooner that later we'll have to refactor the NodeManager registry, at least to move them out of the same lock. I think that will be also a good chance to revisit the individual registries.

Reply:
BTW I think my initial comment still stands `!matches!(e.kind(), _kind)` is never going to be true, see [this StackOverflow question](https://stackoverflow.com/questions/68155018/rust-matches-macro-behaves-unexpectedly-in-regards-to-enum).

Reply:
argh! let me check that, it's actually the first time I felt convinced to use that macro, but guess I'll have to get back to the classic match. Thanks for the heads up!

Reply:
should be ready now 🤞